skimage/segmentation/trainable_segmentation.py

```diff
@@ -0,0 +1,291 @@
+from itertools import combinations_with_replacement
+import itertools
+import numpy as np
+from skimage import filters, feature
+from skimage import img_as_float32
+from joblib import Parallel, delayed
+
+try:
+    from sklearn.exceptions import NotFittedError
+    has_sklearn = True
+except ImportError:
+    has_sklearn = False
+
+    class NotFittedError(Exception):
+        pass
+
+
+
+def _texture_filter(gaussian_filtered):
+    H_elems = [
+            np.gradient(np.gradient(gaussian_filtered)[ax0], axis=ax1)
+            for ax0, ax1 in combinations_with_replacement(range(gaussian_filtered.ndim), 2)
+        ]
+    eigvals = feature.hessian_matrix_eigvals(H_elems)
+    return eigvals
+
+
+
+def _mutiscale_basic_features_singlechannel(
+    img, intensity=True, edges=True, texture=True, sigma_min=0.5, sigma_max=16
+):
+    """Features for a single channel nd image.
+
+    Parameters
+    ----------
+    """
+    # computations are faster as float32
+    img = np.ascontiguousarray(img_as_float32(img))
+    sigmas = np.logspace(
+        np.log2(sigma_min),
+        np.log2(sigma_max),
+        num=int(np.log2(sigma_max) - np.log2(sigma_min) + 1),
+        base=2,
+        endpoint=True,
+    )
+    all_filtered = Parallel(n_jobs=-1, prefer='threads')(delayed(filters.gaussian)(img, sigma) for sigma in sigmas)
+    features = []
+    if intensity:
+        features += all_filtered
+    if edges:
+        all_edges = Parallel(n_jobs=-1, prefer='threads')(delayed(filters.sobel)(filtered_img)
+                                for filtered_img in all_filtered)
+        features += all_edges
+    if texture:
+        all_texture = Parallel(n_jobs=-1, prefer='threads')(delayed(_texture_filter)(filtered_img)
+                                for filtered_img in all_filtered)
+        features += itertools.chain.from_iterable(all_texture)
```

Comment:
I've just been playing around with your trainable segmentation and I like it.
Just wanted to point out that you can avoid the dependency on `joblib.Parallel` (and arguably end up with a more concise code) if you use `ThreadPoolExecutor` from `concurrent.futures` like this:

```
    with ThreadPoolExecutor() as ex:
        all_filtered = list(ex.map(lambda s: filters.gaussian(img,s), sigmas))
        features = []
        if intensity:
            features += all_filtered
        if edges:
            all_edges = list(ex.map(filters.sobel, all_filtered))
            features += all_edges
        if texture:
            all_texture = list(ex.map(_texture_filter, all_filtered))
            features += itertools.chain.from_iterable(all_texture)
        return features
```

However, I don't know what the policy on starting threads within skimage is (if there is one, @jni)

Reply:
One place we do use threading currently is in the `cycle_spin` function. [The implementation there is based on dask](https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/_cycle_spin.py#L141-L144). Dask is currently one of our optional dependencies, so the function has an equivalent single-threaded code path that is used when dask is not present.

Note that for that function, our default for `num_workers` is to use all available cores.

I think I did originally do the cycle_spin implementation using `concurrent.futures` (see #2647), but we switched to `dask` during code review. One advantage of Dask is that you can use different schedulers ('single-threaded', 'threads', 'processes') or even distributed across a number of worker nodes. Note that we do not specify the scheduler in the call to `computer` in the `cycle_spin` example, so the default threaded scheduler unless the user has [configured a different scheduler](https://docs.dask.org/en/latest/scheduling.html) prior to calling the function.

Reply:
One change relative to the time of the older `cycle_spin` discussion is that we have now dropped support for older python that didn't have `concurrent.futures` in the standard library.

Reply:
thanks for the feedback here, I'll try both `concurrent.futures` and `dask` here. Given the fact that `dask` is only an optional dep, maybe it's safer to use `concurrent.futures`? I need to try it first anyway.

Reply:
I played with dask as well this afternoon (local time) and used `dask.bag` to come up with this.

```
    import dask.bag as db
    ...
    sigmas_db = db.from_sequence(sigmas)
    all_filtered = sigmas_db.map(lambda s: filters.gaussian(img,s)).compute(scheduler='threads')
    features = []
    if intensity:
        features += all_filtered
    if edges:
        all_edges = db.from_sequence(all_filtered).map(filters.sobel).compute(scheduler='threads')
        features += all_edges
    if texture:
        all_texture = db.from_sequence(all_filtered).map(_texture_filter).compute(scheduler='threads')
        features += itertools.chain.from_iterable(all_texture)
    return features
```

Running this, I had the subjective feeling that it was slower than the `concurrent.futures` example.
When trying to benchmark this on a similar toy example, I could only confirm a very minor difference in speed, [see here](https://github.com/VolkerH/PythonSnippets/blob/master/daskbag_vs_ThreadPoolExecutor/daskbag_vs_threadpoolexecutor.ipynb).

Off topic, but given that @grlee77 commented... the original reason I looked at this was to see whether this could be quickly ported to `cupyimg` (for the features) and `cuml` (for the RandomForestClassifier). I guess it should be straight forward.

However, `dask.bag` required installation of an additional dependency `fsspec`.

Reply:
> Off topic, but given that @grlee77 commented... the original reason I looked at this was to see whether this could be quickly ported to cupyimg (for the features) and cuml (for the RandomForestClassifier). I guess it should be straight forward.

Yes, not all features have been ported to `cupyimg`, but it looks like the ones needed here have been. In that case it should be pretty straightforward to try!